### PR TITLE
🌱 Clusterctl now logs when objects aren't moved

### PR DESCRIPTION
**Pull request by jichenjc**
**What this PR does / why we need it**:

add a virtual node checker , so that we can remove the TODO in mover function

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by jichenjc** [comment]
not sure this is the right way to go.. just post for some comments

File: cmd/clusterctl/client/cluster/objectgraph.go
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/objectgraph.go:478]: What about
```suggestion
			return errors.Errorf("node %v/%v wont be moved because not included in GVK considered for move", node.identity.Kind, node.identity.Name)
```

File: cmd/clusterctl/client/cluster/mover.go
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/mover.go:85]: I'm not really sure we should error and block in this case, what about logging only?

The main reason to log only is that this check should fail only if there are owner references pointing to objects which are not included in the list considered by CAPI, which is a violation of the clusterctl contract. So, given that, I prefer the user to be informed, but not blocked by clusterctl.

Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/mover.go:85]: @wfernandes opinions? ^^

File: cmd/clusterctl/client/cluster/objectgraph.go
Comment (vincepri) [inline cmd/clusterctl/client/cluster/objectgraph.go:475]: The comment on this method doesn't match the code, what should we actually do? Log an info, or return a set of errors?

**Review by wfernandes — COMMENTED** [review]
Thanks for this PR. 
Could we have a test for this additional logic? I think that may also help document what behavior we are expecting from here.

File: cmd/clusterctl/client/cluster/mover.go
Comment (jichenjc) [reply cmd/clusterctl/client/cluster/mover.go:85]: ok, make sense to me ,will update

File: cmd/clusterctl/client/cluster/objectgraph.go
Comment (jichenjc) [reply cmd/clusterctl/client/cluster/objectgraph.go:475]: should log warning based on above discussion :)
that's original my thought and forgot to update the comment

**Comment by jichenjc** [comment]
> 
> 
> Thanks for this PR.
> Could we have a test for this additional logic? I think that may also help document what behavior we are expecting from here.

I updated the code so that it will not block further actions if it's virtual node.. just log it 
do we still need test case ?

File: cmd/clusterctl/client/cluster/objectgraph.go
Comment (wfernandes) [inline cmd/clusterctl/client/cluster/objectgraph.go:479]: We can make this a level 1 log since it provides more info about the internal workflow of what's happening and it is not critical that the user needs to know about.

I'm referencing https://github.com/kubernetes-sigs/cluster-api/blob/master/cmd/clusterctl/log/doc.go as guidance.

```suggestion
			log.V(1).Info("Object won't be moved because it's not included in GVK considered for move", "kind", node.identity.Kind, "name", node.identity.Name)
```

File: cmd/clusterctl/client/cluster/objectgraph.go
Comment (wfernandes) [inline cmd/clusterctl/client/cluster/objectgraph.go:474]:
```suggestion
// checkVirtualNode logs if nodes are still virtual
```

**Review by wfernandes — COMMENTED** [review]
Since we are just logging I don't see a need for a test. 

However, I'm failing to see _why this is needed?_ 
Is there some benefit to the user to see the logs for the virtual nodes that aren't going to be moved? I believe there was an issue for a dry-run move command. Would that fulfill this need as well?

I see the intent of this PR is to resolve the TODO regarding _ensuring the object graph is complete (no virtual nodes left)_. Do we feel that this PR resolves that? I'm just asking to better understand what we are trying to fix here. Thanks.

@fabriziopandini @vincepri Thoughts?

**Comment by fabriziopandini** [comment]
@wfernandes 
the fact there are still virtual node IMO could be a useful signal for developers that specific implementation (and thus a specific object graph) is not respecting the contract

**Comment by vincepri** [comment]
I'd agree that having logs is usually helpful to either debug or catch potential issues

**Comment by wfernandes** [comment]
@fabriziopandini @vincepri Thanks for the context.
@jichenjc I'm lgtm pending the last couple of nits. Thanks!

File: cmd/clusterctl/client/cluster/objectgraph.go
Comment (jichenjc) [reply cmd/clusterctl/client/cluster/objectgraph.go:479]: done, thanks

File: cmd/clusterctl/client/cluster/objectgraph.go
Comment (jichenjc) [reply cmd/clusterctl/client/cluster/objectgraph.go:474]: done

**Comment by jichenjc** [comment]
@fabriziopandini @vincepri @wfernandes thanks for the comments, I updated accordingly

**Comment by vincepri** [comment]
/milestone v0.3.10
/retitle 🌱 Clusterctl now logs when objects aren't moved

File: cmd/clusterctl/client/cluster/objectgraph.go
Comment (vincepri) [inline cmd/clusterctl/client/cluster/objectgraph.go:479]: Before we merge this, trying to read through this message as a newcomer could be confusing, is there a better way to describe, or some links to documentation on how to debug?

Comment (jichenjc) [reply cmd/clusterctl/client/cluster/objectgraph.go:479]: I guess we don't have a doc about this detail yet
should I write some info (with the information obtained above) then point it to there?

Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/objectgraph.go:479]: As said in a previous comment, IMO this message is mostly for developers
Referece documentation for move/dicovery is https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#move

Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/objectgraph.go:479]: An option is to change log level to v(5) so this is is debugging information as per https://github.com/kubernetes-sigs/cluster-api/blob/master/cmd/clusterctl/log/doc.go

**Comment by jichenjc** [comment]
@fabriziopandini updated from v1 to v5 now.. thanks

**Comment by fabriziopandini** [comment]
@jichenjc thanks!
@vincepri @wfernandes over to you for the final lgtm

**Review by vincepri — APPROVED** [review]
/lgtm

**Pull request merged**
